src/mmu/timers.rs

```diff
@@ -0,0 +1,292 @@
+    pub system_clock_counter: u16,
+    tima_overflowed: bool,
+    tima_write_lock_counter: u16,
+    tima_falling_edge_detected: bool,
+}
+
+impl Timers {
+    pub fn new() -> Self {
+        Timers {
+            system_clock: 0,
+            system_clock_prev: 0,
+            system_clock_counter: 0,
+            tima_overflowed: false,
+            tima_write_lock_counter: 0,
+            tima_falling_edge_detected: false,
+        }
+    }
+}
+
+impl Mmu {
+    // One tick is 1 t-cycle
+    pub fn tick_timers(&mut self) {
+        self.timers.system_clock_counter += 1;
+        self.timers.tima_write_lock_counter = self.timers.tima_write_lock_counter.saturating_sub(1);
+
+        // The system clock is a 16-bit number in t-cycles, but it is only incremented once every m-cycle.
+        if self.timers.system_clock_counter == T_CYCLES_PER_M_CYCLE {
+            self.timers.system_clock_counter = 0;
+            self.timers.system_clock = self.timers.system_clock.wrapping_add(T_CYCLES_PER_M_CYCLE);
+        }
+
+        // ----- Updating DIV
+        let div_value = (self.timers.system_clock >> 8) as u8;
+        self.write_byte_override(DIV_ADDR, div_value);
+
+        // ----- Updating TIMA
+        let tac_enable = self.get_tac_enable();
+        let tima_bit_is_active = self.get_tima_bit_state(false);
+        let tima_bit_was_active = self.get_tima_bit_state(true);
+
+        // TIMA's special overflow behavior occurs the cycle AFTER an overflow is detected.
+        if self.timers.tima_overflowed {
+            self.process_tima_overflow();
+        }
+
+        // TIMA increments while enabled, and on a falling clock edge. The second condition is
+        // there because memory writes can trigger a clock edge detection that occurs elsewhere.
+        if tac_enable && tima_bit_was_active && !tima_bit_is_active
+            || self.timers.tima_falling_edge_detected
+        {
+            self.increment_tima();
+            self.timers.tima_falling_edge_detected = false;
+        }
+
+        // It's important to update the previous clockstate here, instead of at the beginning of the loop.
+        // This is because the system clock can be reset if something writes to the div timer.
+        // Therefore, the system clock state might be different by the next time this function is called.
+        self.timers.system_clock_prev = self.timers.system_clock;
+    }
+
+    fn increment_tima(&mut self) {
+        let mut tima_byte = self.read_byte(TIMA_ADDR);
+        let overflowed;
+
+        (tima_byte, overflowed) = tima_byte.overflowing_add(1);
+
+        self.write_byte_override(TIMA_ADDR, tima_byte);
+
+        if overflowed {
+            self.timers.tima_overflowed = true;
+        }
+    }
+
+    /// Overflows of TIMA send a timer interrupt. When TIMA overflows, its value is set
+    /// to the value stored in TMA.
+    fn process_tima_overflow(&mut self) {
+        self.timers.tima_overflowed = false;
```

Comment:
I'm not saying you need to do it here, but this is an opportunity to write some defensive code.  In this case, you might want to make sure that TIMA really overflowed before resetting it.  You can use

```
debug_assert!(self.timers.tima_overflowed);
self.timers.tima_overflowed = false;
```

Like assert!() it will panic if the assertion if false. Unlike assert!(), it will be compiled out of the release code so it won't slow you down or cause a panic in release.